org.eclipse.jdt.apt.core/src/org/eclipse/jdt/apt/core/internal/env/BinaryFileOutputStream.java

```diff
@@ -46,40 +43,20 @@ public BinaryFileOutputStream(IFile file, BuildEnv env) {
 	public void close() throws IOException {
 		super.close();
 
-		InputStream contents = new ByteArrayInputStream(toByteArray());
+		byte[] newContent = toByteArray();
+		boolean contentsChanged = true;
 		try {
-
-			boolean contentsChanged = true;
-			if (!_file.exists()) {
-				saveToDisk(contents, true);
-			}
-			else {
-				InputStream in = null;
-				InputStream oldData = null;
-				try {
-					// Only write the contents if the data is different
-					in = new ByteArrayInputStream(toByteArray());
-					oldData = new BufferedInputStream(_file.getContents());
-					if (FileSystemUtil.compareStreams(in, oldData)) {
-						contentsChanged = false;
-					}
-				}
-				catch (CoreException ce) {
-					// Ignore -- couldn't read the old data, so assume it's different
-					contentsChanged = true;
-				}
-				finally {
-					closeInputStream(in);
-					closeInputStream(oldData);
-				}
-				if (contentsChanged) {
-					contents.reset();
-					saveToDisk(contents, false);
-				}
+			// Only write the contents if the data is different
+			byte[] oldContent = _file.readAllBytes();
```

Comment:
I suppose it's a trade-off to avoid an exception by checking `_file.exists()` versus the extra cost of the test itself for the very common case that the target already exists, right?

Reply:
org.eclipse.jdt.internal.apt.pluggable.core.filer.IdeNonSourceOutputStream,
org.eclipse.jdt.internal.apt.pluggable.core.filer.IdeClassOutputStream,
org.eclipse.jdt.apt.core.internal.env.BinaryFileOutputStream,

are now refactored to use common FileSystemUtil.saveToDisk(IFile, byte[])

Reply:
> I suppose it's a trade-off to avoid an exception by checking _file.exists() versus the extra cost of the test itself for the very common case that the target already exists, right?

yes, readAllBytes(); will throw CoreException if the file does not exists, which will result in the common case of saveToDisk because contentsChanged. so there is not need for an extra fie.exists() check.